### PR TITLE
Add setting for audio sound sources limit

### PR DESCRIPTION
**Pull request by doug65536**
Add compact, short information about your PR for easier understanding:

I noticed a lot of warnings in the sound code, saying createPlayingSound had out of memory. I have a lot of memory and quickly found that it was the error code for hitting the limit on sound sources.
I figured out how to add the setting and updated the sound initialization code.
Surprisingly, nobody reported a bug for flooding warnings from createPlayingSound saying out of memory.
It fixes a flood of sound warnings when you are near a farm that is making a lot of sounds. There are no warnings with it set to 1024 for my farm. A much lower sound would work too. I checked `perf top`, and I can't even see sound related code in the profile.

If 128 sounds is an acceptable default number of sound sources, you can consider this ready to go. If you would prefer it preserve the previous behaviour as much as possible, I can make the setting "0" mean "default", and have it use a nullptr attribute list parameter, and only use the attribute list if the setting was not zero.

## How to test
I added a setting in the Audio page.

### CODE REVIEW

**Comment by Desour** [comment]
I could reproduce getting warnings, with `/lua for i = 1, 300 do core.sound_play("<some stereo sound>", {}) end`.

If we do this, we should also set `ALC_MONO_SOURCES` accordingly.
To cite OpenAL spec chapter 6.2.1. Context Attributes:
> `ALC_MONO_SOURCES`
> A hint indicating how many sources should
be capable of supporting mono data
> `ALC_STEREO_SOURCES`
> A hint indicating how many sources should
be capable of supporting stereo data

However, I'm not sure if we should set these attributes. There's a setting in `alsoft-config`, at "Resources" → "Number of Sound Sources" (in alsoft.conf: `sources` under `[General]`), which is 256 by default. Users can just modify that value instead.

Also, increasing this limit doesn't fix the actual issue. Playing so many sounds at once performs really bad, at least on my machine.
Maybe we should implement a source limit per sound name.

**Comment by doug65536** [comment]
I am curious what you did to cause a performance problem. Do you mean you made a for loop that explicitly starts 1024 (or some huge number) of sounds all at once?

I am curious because, I was afraid of performance issues, and it breezed through my in-game mineclone2 subterranean farm that is jam packed with cows and chickens and pigs. On defaults, going near that completely flooded output with createPlayingSound yellow warnings that say out of memory (which means out of sources in this context). On 1024 sources, I could not even see one sound related thing in `perf top`, and zero warnings, and you could hear the difference, it made more sense.

I measured on a headset that would cause openal to mix to stereo channels, and send them to USB PCM stream, so it should have been a pathological scenario. If I let my "motherboard" audio do it, with the CPU's fancy SoC audio hardware, it probably can do hardware acceleration. I tested on pathological sound device, a USB headset. I couldn't even see an audio function in the profile.

Starting 1024 sounds all at once is a bit of an unrealistic scenario, a strawman, which artificially inflates the cost of having it set to a large number of sources, by putting the setup overhead of the entire limit together end to end, and arguing that the sum is too much. It's like saying we can't power a city, because if everyone turned on a vacuum at the same time it is too much current. It is far more expensive to start a sound, than it is to play a sound. If they are naturally staggered out over time, where they start at different times, it can handle doing a bit of mixing on a large number of channels.

**Comment by Desour** [comment]
>Do you mean you made a for loop that explicitly starts 1024 (or some huge number) of sounds all at once?

I did both, many at once and multiple times only 20 at once.

`perf top` shows `libopenal.so` at the top for me. And a cpu core is fully utilized.

>It is far more expensive to start a sound, than it is to play a sound.

I also used a sound that lasts a little longer (multiple seconds), so that multiple buffers are enqueued over time.
To rule out the difference, I've also tried using short stereo sounds (here the pickup sound from `item_drop`):
`/lua local function fun() for i = 1, 5 do core.sound_play("item_drop_pickup_1", {}) end core.after(0, fun) end fun()`

And I've tried my laptop's builtin speakers, headphones with bluetooth and headphones with headphone jack. I got everywhere about the same results.

I'm using OpenAL-soft 1.23.1, pulseaudio and linux btw..

---

Anyway, I'm not saying you should leave that setting at default. But you do not need minetest to do it if you can use alsoft-config instead.
(Idk about availability of such settings in other OpenAL implementations.)

**Comment by SmallJoker** [comment]
Isn't it simply a mod issue when too many sounds are played at once? Changing these limits would only hide the problem of unoptimized mods.

**Comment by rubenwardy** [comment]
> Isn't it simply a mod issue when too many sounds are played at once? Changing these limits would only hide the problem of unoptimized mods.

No, multiple mods can be creating sounds at once, the number of sounds also varies system to ststem. The engine should automatically prioritise sounds and only play the most relevant (ie: the loudest)

**Comment by Desour** [comment]
>Also, rather than having a setting for sound sources limit it would be good to ask OpenAL / the system instead, if that is possible

That should be equivalent to not explicitly setting the `ALC_STEREO_SOURCES` hint.

---

Considering all of the points above, and because we don't set any other alsoft-config configurable setting either yet, like forced hrtf, I think we should reject this PR.

**Comment by doug65536** [comment]
If we don't set it, we are just throwing the system default at the user. Who has even heard of tweaking their audio API defaults in some system config file? What is it on windows, regedit? Come on. Games should not need you to be a sysadmin. There should be a setting where they can put the audio on "default" (perhaps doing what it does now - not even set it), "low" or "high", bare minimum.

I still can't repro it using a lot of CPU. I ended up going as far as creating a mod that implements `/lua` and I ran into the sound name being not found. I did find a place in-game where crouching in the water sprays you with a really large number of splashy sounds at once, and I still can't get openal to even show up in `perf top`, with zeroing enabled, on my 3950x, outputting to a USB headset that can't hardware mix, it's just playing a software mixed stereo PCM stream.

**Comment by doug65536** [comment]
Would it simultaneously please everyone if it could be default, low, high, or custom, and have the next field editable if custom, readonly if low or high.
- Default does what it does now,
- Otherwise it uses the custom number,
- Selecting low or high sets the "custom" setting to hardcoded numbers for low and high, like 32 and 128 or whatever
- And setting it to 32 or 128 flicks the other magically to low or high
- And setting it to something else flicks it to custom

Then it's no change until someone goes and pokes at the settings. And there is a setting where you can low-spec or crank it up, according to user intent. When I saw the log flood saying that there were too many sounds, I would have gone and put it on high and it would have been fixed, and I would have a fun huge number of animal sounds simultaneously on my big desktop, without being a sysadmin.

**Comment by doug65536** [comment]
Maybe even, when it is on default, query what it ended up using, and set the setting field to that number, disabled. So you can conveniently see what "default" is using for the voice limit.

**Comment by Desour** [comment]
>I ended up going as far as creating a mod that implements /lua and I ran into the sound name being not found.

FYI, if you play sounds, the sound name is the sound file name without the ending (`.ogg` or `.<num>.ogg`), see doc.
And the lua command comes from the luacmd mod. Worldedit also has a similar command.

---

>If we don't set it, we are just throwing the system default at the user. Who has even heard of tweaking their audio API defaults in some system config file? What is it on windows, regedit? Come on. Games should not need you to be a sysadmin.

`alsoft-config` is a gui program actually (just without a desktop entry). Idk what it is like on windows, or on android, or mac.

But that's actually a fair point. I shouldn't judge ease of setting system defaults just by my own system.
And I'm not per se against adding advanced sound settings. We wanted to add advanced settings, e.g. to set hrtf, anyway, iirc. So, let's just start with this setting.

Btw., I've also looked at other games and applications using openal. The only one I found that has a similar setting was openspades, a "polyphonics" setting, but it's 96 by default and doesn't go higher than 256. To be fair, many apps don't do things too well, or they don't have the same situation like minetest where there can easily be an arbitrary number of sounds.

>default, low, high, or custom

That would just over-complicate things.
Just add a setting for the max number of sound sources, and don't set (aka use system default) if it's a special value, i.e. -1.

**TL;DR:** I changed my opinion, and give concept approval. More PRs for advanced sound settings can follow afterwards, this is the first.

Things to change in this PR:
* Also set the `ALC_MONO_SOURCES` hint. Either by a second setting or the same, I don't care.
* Don't set it if the setting is -1.
* Update `defaultsettings.cpp`. (Engine settings have defaults.) (Default should be -1.)
* Rebase. (`SoundManagerSingleton` is now in `src/client/sound/sound_singleton.cpp`.)

**Comment by doug65536** [comment]
I don't know what you did to git. I can't get it to accept my changes cleanly. I should probably give up.

**Comment by doug65536** [comment]
My fork is in shambles now I guess.

**Comment by Desour** [comment]
> My fork is in shambles now I guess.

FYI, normally, you'd use a feature branch, and force push after a rebase. This might be helpful: https://forum.minetest.net/viewtopic.php?f=3&t=14262&p=211498#p211502
